api/analyzedphenotypes.genus.api.inc

```diff
@@ -0,0 +1,194 @@
+    ),
+    'assign_genus' => array(
+      'func'  => 'ap_set_projectgenus',
+      'param' => array('project_id' => $parameter['project_id'], 'genus' => $parameter['genus']),
+      'opts'  => null,
+    ),
+    'sysvar_genus_set' => array(
+      'func'  => 'ap_get_activegenus',
+      'param' => null,
+      'opts'  => null,
+    ),
+  );
+
+  // Function name, parameter (project name key or name), additional options.
+  return call_user_func($command[$property]['func'], $command[$property]['param'], $command[$property]['opts']);
+}
+
+
+///////
+
+
+/**
+ * FUNCTION:
+ * Assign a genus to an experiment/project in chado.projectprop table.
+ * Dependencies: ap_get_variablenames() - systemvars.api.
+ *
+ * @param $identifiers
+ *   An array with the following keys:
+ *   - project_id : project id number, to which the genus will be linked to.
+ *   - genus : genus.
+ */
+function ap_set_projectgenus($identifiers) {
+  // Fetch the configuration for the term Genus from terms system variable set.
+  $sysvar = ap_get_variablenames(
+    array('variablename' => 'genus'),
+    array('set' => 'terms')
+  );
+
+  $sysvar_genus = variable_get($sysvar);
+
+  // Before relating genus to a project, test for any previous correlations.
+  $is_rel = chado_query(
+    "SELECT projectprop_id FROM {projectprop} WHERE value = :genus AND project_id = :project_id AND type_id = :sysvar_genus",
+    array(
+      ':genus'   => $identifiers['genus'],
+      ':project_id' => $identifiers['project_id'],
+      ':sysvar_genus' => $sysvar_genus,
+    )
+  )
+  ->fetchCol();
+
+  if (!$is_rel) {
+    // Is clear, no past relations made.
+    chado_insert_record('projectprop', array(
+      'project_id' => $identifiers['project_id'],
+      'type_id' => $sysvar_genus,
+      'value' => $identifiers['genus']
+    ));
+  }
+}
+
+/**
+ * FUNCTION:
+ * Fetch genus assigned to an experiment/project in chado.projectprop table.
+ * Dependencies: ap_get_variablenames() - systemvars.api.
+ *
+ * @param $identifiers
+ *   An array, with the following keys:
+ *     - value : project id number or project name.
+ *
+ * @param $options
+ *   An array to be used as additional options to be performed to
+ *   returned result.
+ *   - is_id : boolean, to flag if supplied value is a project id or not. Default to TRUE.
+ *
+ *  @return
+ *    A string, genus of an experiment/project.
+ */
+function ap_get_projectgenus($identifiers, $options = array('is_id' => TRUE)) {
+  // Fetch the configuration for the term Genus from terms system variable set.
+  $sysvar = ap_get_variablenames(
+    array('variablename' => 'genus'),
+    array('set' => 'terms')
+  );
+
+  $sysvar_genus = variable_get($sysvar);
+
+  // What is provided? project id or project name?
+  $field = ($options['is_id']) ? 'project.project_id' : 'project.name';
+
+  // Fetch projectprop.value of type (type_id) Genus for a project id or project name.
```

Comment:
This can be made more efficient by having separate SQL statements depending upon what is passed in. Joins are expensive so don't make them needlessly.
```php
if (isset($identifiers['project_id'])) {
    // SELECT value FROM {projectprop} WHERE project_id=:value AND type_id=:type
}
elseif (isset($identifiers['name'])) {
  // Your current query.
}
```